File: algorithm/236.二叉树的最近公共祖先.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 * 
 * 思路1:
 *  这个是找普通二叉树的公共祖先节点
 *  可以用往下遍历的方法 找到 p 和 q 的所有父亲节点的链路
 *  之后再比较从 root 节点到它们自己本身的这两条链路上的最后一个相同的节点就可以了
 */
var lowestCommonAncestor = function(root, p, q) {
  const _findNodeFater = (root, target) => {
    const res = [];
    if (!root || !target) return res;
    let currentNode = root;
    let prevNode = null;
    while(true) {
      if (currentNode.val === target.val) {
        res.push(currentNode);
        return res;
      }
      if (currentNode.left) {
        res.push(currentNode);
        currentNode = currentNode.left;
      } else if (currentNode.right) {
        res.push(currentNode);
        currentNode = currentNode.right;
      } else {
        while(!currentNode.right || (currentNode.right === prevNode)) {
          prevNode = currentNode;
          currentNode = res.pop();
          if (!currentNode) return res;
        }
        res.push(currentNode);
        currentNode = currentNode.right;
      }
    }
  }

  const pFater = _findNodeFater(root, p);
  const qFater = _findNodeFater(root, q);
  for (let i = pFater.length; i--;) {
    const pi = pFater[i].val;
    for (let j = qFater.length; j--;) {
      const qj = qFater[j].val;
      if (pi === qj) return pFater[j];
    } 
  }
};

File: algorithm/236.二叉树的最近公共祖先.cpp
#include "iostream"
#include "string"
#include "vector"
using namespace std;

/**
 * 思路2:
 *  递归
 *  如果没有 root 说明是找到了叶子节点
 *  如果发现等于 p 或者 q 则说明找到了目标节点 直接返回
 *  在该层递归的上一层判断
 *    1. 如果 左右节点均有值 说明找到了公共祖先节点
 *    2. 如果 左右两棵子树只有一个有值的话说明当前左或右子树有值 直接返回
 *    3. 如果 都是 null 则直接返回 null
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if (!root) return nullptr;
    if (root == p || root == q) {
      return root;
    }
    TreeNode* l = lowestCommonAncestor(root -> left, p, q);
    TreeNode* r = lowestCommonAncestor(root -> right, p, q);
    if (l && r) {
      return root;
    }
    return r ? r : l;
  }
};